scanner/src/main/kotlin/experimental/ScanController.kt

```diff
@@ -180,6 +180,20 @@ class ScanController(
             nestedProvenance != null && getMissingProvenanceScans(scanner, nestedProvenance) == listOf(provenance)
         }.map { (id, _) -> packages.first { it.id == id } }
 
+    /**
+     * Return a map of [KnownProvenance]s associated with all [packages] with the same provenance, ignoring any VCS
+     * path. Packages without a resolved provenance are not included in the result.
+     */
+    fun getPackagesConsolidatedByProvenance(): Map<KnownProvenance, List<Package>> =
+        packages.mapNotNull { pkg ->
+            packageProvenances[pkg.id]?.let { pkg to it }
+        }.groupBy { (_, provenance) ->
+            when (provenance) {
+                is RepositoryProvenance -> provenance.copy(vcsInfo = provenance.vcsInfo.copy(path = ""))
+                else -> provenance
+            }
+        }.mapValues { it.value.map { (pkg, _) -> pkg } }
```

Comment:
Alternatively (using only a single loop):

    fun getPackagesConsolidatedByProvenance(): Map<KnownProvenance, List<Package>> {
        val packagesByProvenance = mutableMapOf<KnownProvenance, MutableList<Package>>()

        packages.forEach { pkg ->
            val consolidatedProvenance = when (val provenance = packageProvenances[pkg.id]) {
                null -> return@forEach
                is RepositoryProvenance -> provenance.copy(vcsInfo = provenance.vcsInfo.copy(path = ""))
                else -> provenance
            }

            packagesByProvenance.getOrPut(consolidatedProvenance) { mutableListOf() } += pkg
        }

        return packagesByProvenance
    }